### PR TITLE
feat(plugin): add ICP token creation support

### PR DESCRIPTION
**Pull request by asDNSk**
- Add plugin-icp to enable token creation on Pickpump platform
- Implement createToken functionality for Internet Computer Protocol (ICP)
- Support ICP token standard and specifications
- Enable seamless integration with Pickpump's token creation flow

BREAKING CHANGE: None

### CODE REVIEW

**Comment by lalalune** [comment]
this is fucking rad, will review later

File: packages/plugin-icp/src/actions/createToken.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    type Action,
+} from "@ai16z/eliza/src/types";
+import { idlFactory } from "../canisters/pick-pump/index.did";
+import { _SERVICE } from "../canisters/pick-pump/index.did.d";
+import { ActorCreator, CreateMemeTokenArg } from "../types";
+import { unwrapOption, wrapOption } from "../utils/common/types/options";
+import { unwrapRustResultMap } from "../utils/common/types/results";
+import { icpWalletProvider } from "../providers/wallet";
+
+const createTokenTemplate = `Respond with a JSON markdown block containing only the extracted values. Use null for any values that cannot be determined.
```
Comment (monilpat) [inline packages/plugin-icp/src/actions/createToken.ts:19]: Can we move this into a templates file for readability? Thanks so much!

File: packages/plugin-icp/src/actions/createToken.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+Generate appropriate token information based on the user's description.
+Respond with a JSON markdown block containing only the generated values.`;
+
+async function createTokenTransaction(
+    creator: ActorCreator,
+    tokenInfo: CreateMemeTokenArg
+): Promise<any> {
+    const actor: _SERVICE = await creator(
+        idlFactory,
+        "bn4fo-iyaaa-aaaap-akp6a-cai"
```
Comment (monilpat) [inline packages/plugin-icp/src/actions/createToken.ts:52]: Can this be moved to a constants file?

File: packages/plugin-icp/src/actions/createToken.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            template: createTokenTemplate,
+        });
+
+        const response = await generateObject({
+            runtime,
+            context: createTokenContext,
+            modelClass: ModelClass.LARGE,
+        });
+
+        console.log("Response:", response);
```
Comment (monilpat) [inline packages/plugin-icp/src/actions/createToken.ts:112]: We can remove this

File: packages/plugin-icp/src/actions/createToken.ts [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            const createTokenResult = await createTokenTransaction(creator, {
+                name: response.name,
+                symbol: response.symbol,
+                description: response.description,
+                website: response.website,
+                twitter: response.twitter,
+                telegram: response.telegram,
+            });
+
+            console.log("Token created successfully:", createTokenResult);
```
Comment (monilpat) [inline packages/plugin-icp/src/actions/createToken.ts:144]: same here

File: packages/plugin-icp/src/types.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import type { Principal } from "@dfinity/principal";
+import type { ActorSubclass } from "@dfinity/agent";
+import type { IDL } from "@dfinity/candid";
+export interface ICPConfig {
```
Comment (monilpat) [inline packages/plugin-icp/src/types.ts:4]: nit: space

File: packages/plugin-icp/src/utils/common/data/json.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { isPrincipalText } from "../../ic/principals";
+// ? 1. bigint -> string
+// ? 2. principal -> string
+
+export const customStringify = (v: any): string =>
+    JSON.stringify(v, (_key, value) => {
+        if (typeof value === "bigint") return `${value}`;
+        if (value && typeof value === "object" && value._isPrincipal === true) {
```
Comment (monilpat) [inline packages/plugin-icp/src/utils/common/data/json.ts:8]: For readability: else if probably more appropriate

File: packages/plugin-icp/src/utils/common/data/price.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import _ from "lodash";
+
+export const handleShowPrice = (value: number | string, fixed = 8) => {
+    if (!value || value === "0") return 0;
+    return _.round(Number(value), fixed);
+    // return value / 1e18;
```
Comment (monilpat) [inline packages/plugin-icp/src/utils/common/data/price.ts:6]: Can we remove this comment

File: packages/plugin-icp/src/utils/hex.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+// hex text -> number array
```
Comment (monilpat) [inline packages/plugin-icp/src/utils/hex.ts:1]: We could consider moving this to the array utils as they are all related to arrays

File: packages/plugin-icp/src/utils/ic/index.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
+import { getCrc32 } from "@dfinity/principal/lib/esm/utils/getCrc";
+import { sha224 } from "@dfinity/principal/lib/esm/utils/sha224";
+
+import { Principal } from "@dfinity/principal";
+import { array2hex, hex2array } from "../hex";
+import { string2array } from "../arrays";
+
+// Principal -> string
+export const principal2string = (p: Principal): string => p.toText();
```
Comment (monilpat) [inline packages/plugin-icp/src/utils/ic/index.ts:9]: Can remove and use util for these

**Review by monilpat — CHANGES_REQUESTED** [review]
This is some amazing work - thank you for working on this! Added a handful of small nits feel free to address and we are good to go!

**Comment by monilpat** [comment]
Also feel free to fix the PR Title check, add in a a screen grab of successful token creation, and consider an integration test to simulate the entire token creation flow using mocks or a test canister to ensure the plugin works end-to-end. Also can we update the plugin associated READMEs so it is easier to leverage. Thanks so much! So sorry for all the comments! You are a badass and did some great work here!

**Comment by asDNSk** [comment]
Thank u for ur review, I will fix the issues. @monilpat

**Comment by NashAiomos** [comment]
Oh yeah

**Pull request merged**